kore/src/Kore/Step/Simplification/Pattern.hs

```diff
@@ -38,6 +41,42 @@ import           Kore.Syntax.Variable
 import           Kore.Unparser
 import           Kore.Variables.Fresh
 
+simplifyAndRemoveTopExists
+    ::  ( Ord variable
+        , Show variable
+        , Unparse variable
+        , FreshVariable variable
+        , SortedVariable variable
+        )
+    => SmtMetadataTools StepperAttributes
+    -> PredicateSimplifier
+    -> TermLikeSimplifier
+    -- ^ Evaluates functions in patterns.
+    -> BuiltinAndAxiomSimplifierMap
+    -- ^ Map from axiom IDs to axiom evaluators
+    -> Pattern variable
+    -> Simplifier (OrPattern variable)
+simplifyAndRemoveTopExists
+    tools
+    substitutionSimplifier
+    termSimplifier
+    axiomIdToSimplifier
+    patt
+  = do
+    simplified <-
+        simplify
+            tools
+            substitutionSimplifier
+            termSimplifier
+            axiomIdToSimplifier
+            patt
+    return (removeTopExists <$> simplified)
```

Comment:
Should we remove the `\exists` before or after simplification? `\exists` simplification can leave predicates in the term, so could they be left in the term here?

Reply:
I asked myself the same question. I'm not sure, I think that both are somewhat broken:

* If we want to remove the existentials for rewrite rules that we just added in this PR, then we can remove them both at the beginning and at the end, and removing them at the beginning works better, since we keep the term-predicate-substitution split.

* If we want to remove other existentials, then we need to lift them. In principle we should be able to lift and remove them before simplification, but we're not using quantifiers in a consistent way (yet), so we're going to break things. Here's the issue:

In symbolic execution we're usually trying to prove a `phi->psi` formula. The symbolic variables are quantified universally, i.e. `forall x . (phi -> psi)`, and we don't have explicit quantifiers for them. The new variables are quantified existentially in `phi`, i.e. `(exists y . phi)->psi`, which can also be written as `forall y . (phi -> psi)`, which means that they are similar enough to the symbolic variables to remove quantifiers. But the problem is that, sometimes, we prefer to have them as existential quantifiers, since `exists x . (phi and [x=alpha])`, where `x` does not occur in `alpha`, can be transformed into a formula that does not contain `x` at all, and we use that for `owise` clauses. But if we remove the quantifier, i.e. we transform the variable in a symbolic one, we'll never be able to fully get rid of it.

So, maybe a better solution would have been to make unification work with quantifiers (at least with the top ones) and to not remove them, but that would have made everything somewhat more complicated. So, short-term, removing exists before simplification is better, since currently we don't lift quantifiers. Medium term, removing it after simplification is better. Long-term, there's a good chance that both are bad. I just picked the medium term solution.